pkg/storage/tsdb/index/chunk.go

```diff
@@ -15,3 +19,58 @@ type ChunkMeta struct {
 
 	Entries uint32
 }
+
+type ChunkMetas []ChunkMeta
+
+func (c ChunkMetas) Len() int      { return len(c) }
+func (c ChunkMetas) Swap(i, j int) { c[i], c[j] = c[j], c[i] }
+
+// Sort by (MinTime, MaxTime, Checksum)
+func (c ChunkMetas) Less(i, j int) bool {
+	a, b := c[i], c[j]
+	if a.MinTime != b.MinTime {
+		return a.MinTime < b.MinTime
+	}
+
+	if a.MaxTime != b.MaxTime {
+		return a.MaxTime < b.MaxTime
+	}
+
+	return a.Checksum < b.Checksum
+}
+
+// finalize sorts and dedupes
+// TODO(owen-d): can we remove the need for this by ensuring we only push
+// in order and without duplicates?
+func (c ChunkMetas) finalize() ChunkMetas {
+	sort.Sort(c)
+
+	if len(c) == 0 {
+		return c
+	}
+
+	var res ChunkMetas
+	lastDuplicate := -1
+	prior := c[0]
+
+	// minimize reslicing costs due to duplicates
+	for i := 1; i < len(c); i++ {
+		x := c[i]
+		if x.Checksum == prior.Checksum {
```

Comment:
Should we also check for min/max time to avoid collision in checksum behind considered as duplicates ?

Reply:
Yeah, i think this is probably a good addition for correctness's sake.